arch/x86/core/x86_mmu.c

```diff
@@ -136,3 +136,110 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
 		addr += MMU_PAGE_SIZE;
 	}
 }
+
+#ifdef CONFIG_X86_USERSPACE
+
+/* Helper macros needed to be passed to x86_update_mem_domain_pages */
+#define X86_MEM_DOMAIN_SET_PAGES   (0U)
+#define X86_MEM_DOMAIN_RESET_PAGES (1U)
+/* Pass 1 to reset_pages if reset of mem domain pages is needed else pass a 0*/
+static inline void x86_update_mem_domain_pages(struct k_mem_domain *mem_domain,
+					       u32_t reset_pages)
+{
+	u32_t partition_index;
+	u32_t total_partitions;
+	struct k_mem_partition partition;
+	u32_t partitions_count;
+
+	do {
+		/* If mem_domain doesn't point to a valid location return.*/
+		if (mem_domain == NULL) {
+			break;
+		}
+
+		/* Get the total number of partitions*/
+		total_partitions = mem_domain->num_partitions;
+		__ASSERT(total_partitions <= CONFIG_MAX_DOMAIN_PARTITIONS,
+			 "invalid partitions");
+
+		/* Iterate over all the partitions for the given mem_domain
+		 * For x86: interate over all the partitions and set the
+		 * required flags in the correct MMU page tables.
+		 */
+		partitions_count = 0;
+		for (partition_index = 0;
+		    partitions_count < total_partitions;
+		    partition_index++) {
+
+			/* Get the partition info */
+			partition = mem_domain->partitions[partition_index];
+			if (partition.size == 0) {
+				continue;
+			}
+			partitions_count++;
+			if (reset_pages == X86_MEM_DOMAIN_SET_PAGES) {
+				/* Set the partition attributes */
+				_x86_mmu_set_flags((void *)partition.start,
+						   partition.size,
+						   partition.attr,
+						   K_MEM_PARTITION_PERM_MASK);
+			} else {
+				/* Reset the pages to supervisor RW only */
+				_x86_mmu_set_flags((void *)partition.start,
+						   partition.size,
+						   K_MEM_PARTITION_P_RW_U_NA,
+						   K_MEM_PARTITION_PERM_MASK);
+			}
+		}
+
+	} while (0);
```

Comment:
what is this point of having all this in a do { } while (0) loop?
get rid of this

Reply:
This was placed for MISRA compliance for later on. which states only 1 return per function.

Reply:
don't it this way. use a goto if you want exactly one "return" statement.

Reply:
If I am not mistaken MISRA will complain for the 'goto' as well. So I implemented it this way.

Reply:
We use `goto` a lot in our code; in fact, that's the preferred way of doing error recovery in Zephyr, as that's the one that generates smaller and more efficient code.

The rule in MISRA about the use of this statement (15.1) is only an advisory, and it has an exception to allow jumping only forward (15.2), which is what's used in Zephyr.

Reply:
Thanks for the clarification. Will go ahead with the goto implementation.